### PR TITLE
Finished slice with testing.

### PR DESCRIPTION
**Pull request by preactive**
Was trying to play with yield with fib function in last lesson.  Is there a better way to adjust variables in function with Yield and not use global variables?

Is asserts inline acceptable in the fashion I did them?

### CODE REVIEW

**Comment by preactive** [comment]
I did a `git commit --amend` for PEP8 fixes for slice.py

**Comment by natasha-aleksandrova** [comment]
@preactive `git commit --amend` is appropriate when there is a need for it 🙂 

Typically you would make a new commit, remember that in a way commits tell a story of progression of the code.

example of commit history, most recent at the top
--- fix style for pep8
--- add fizz buzz assignment

when someone else or you a year later trying to figure out why something was added, commit history is a great tool.

amazingly that are even guidelines for commit messages 😃 
https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

File: students/preactive/Lesson2/series.py [hunk 1]
```diff
@@ -7,6 +7,20 @@ def fib(n):
 # print(fib(8))
 
 
+a,b = 0,1
+def fib_gen():
+    global a, b
+    while True:
+        a, b = b, a + b
+        yield a
```
Comment (natasha-aleksandrova) [inline students/preactive/Lesson2/series.py]: as far as your question on yield (btw awesome that you are trying out `yield`)...

you shouldn't need globals here, this works as well:
```
def fib_gen():
    a,b = 0,1
    while True:
        a, b = b, a + b
        yield a
```

File: students/preactive/Lesson3/slice.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    # with the first and last items exchanged
+    swapped = slicer_input[-1:] + slicer_input[1:len(slicer_input) - 1] \
+              + slicer_input[:1]
+    if for_type is str:
+        assert swapped == "ghis is a strint", "String Failed"
+    if for_type is tuple:
+        assert swapped == (32, 54, 13, 12, 5, 2), "Tuple Failed"
+    if for_type is list:
+        assert swapped == ["!", "!", "tide", "pod", "challenge", "is",
+             "not", "for", "the", "weak", "!", "!"], "List failed"
```
Comment (natasha-aleksandrova) [inline students/preactive/Lesson3/slice.py]: For your assert question.. I would keep asserts separate from the actual code.


in real project tests would go at the root of the project in the `tests` folder, in small scripts like this they would go into `if __name__ == '__main__'` block

which would now make your test logic much simpler:

`assert first_last("This is a string") == "ghis is a strint"` without the need for type checking

File: students/preactive/Lesson3/slice.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        first_last(slice_item, for_type)
+        odd_out(slice_item, for_type)
+        ambigus_ao_fourth(slice_item, for_type)
+        reversed(slice_item, for_type)
+        thirds(slice_item, for_type)
+
+
+def first_last(slicer_input, for_type):
+    # with the first and last items exchanged
+    swapped = slicer_input[-1:] + slicer_input[1:len(slicer_input) - 1] \
```
Comment (natasha-aleksandrova) [inline students/preactive/Lesson3/slice.py]: note that the middle part can be simplified to: `slicer_input[1:-1]`

File: students/preactive/Lesson3/slice.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    if for_type is tuple:
+        assert slicer_reversed == (32, 5, 12, 13, 54, 2), "Tuple Failed"
+    if for_type is list:
+        assert slicer_reversed == ['!', '!', 'weak', 'the', 'for', 'not',
+                                   'is', 'challenge', 'pod', 'tide', '!', '!'], \
+            "List failed"
+    return slicer_reversed
+
+
+def thirds(s_i, for_type):
```
Comment (natasha-aleksandrova) [inline students/preactive/Lesson3/slice.py]: more readable to call the var either of these: `seq` or `sequence`

File: students/preactive/Lesson3/slice.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    # with the middle third, then last third, then the
+    # first third in the new order
+    if len(s_i) % 3 == 0:
+        L = round(len(s_i) / 3)
```
Comment (natasha-aleksandrova) [inline students/preactive/Lesson3/slice.py]: var names should be lower case, and avoid one letters (unless comprehensions and other short hand cases)

File: students/preactive/Lesson3/slice.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    a_arr = ["!", "!", "tide", "pod", "challenge", "is",
+             "not", "for", "the", "weak", "!", "!"]
+    three_types = (a_string, a_tuple, a_arr)
+
+    for slice_item in three_types:
+        for_type = type(slice_item)
+        first_last(slice_item, for_type)
+        odd_out(slice_item, for_type)
+        ambigus_ao_fourth(slice_item, for_type)
+        reversed(slice_item, for_type)
```
Comment (natasha-aleksandrova) [inline students/preactive/Lesson3/slice.py]: be careful about overwriting built-in functions, as you can see github highlights `reversed`, so will some code editors.

**Comment by natasha-aleksandrova** [comment]
@preactive I don't believe I see your submission for series assignment in the course website?

Please be sure to submit PR link to get the grade.

File: students/preactive/Lesson2/series.py (hunk 1 above)
Comment (preactive) [reply students/preactive/Lesson2/series.py]: added the test and a modifed one that leads series with 0

File: students/preactive/Lesson3/slice.py (hunk 3 above)
Comment (preactive) [reply students/preactive/Lesson3/slice.py]: Fixed

File: students/preactive/Lesson3/slice.py (hunk 5 above)
Comment (preactive) [reply students/preactive/Lesson3/slice.py]: I was getting lazy.  Thank your for holding me to this.  Fixed.

File: students/preactive/Lesson3/slice.py (hunk 6 above)
Comment (preactive) [reply students/preactive/Lesson3/slice.py]: Fixed, changed verbage

**Comment by preactive** [comment]
I checked the others and I did submit them in the portal

File: students/preactive/Lesson3/slice.py (hunk 4 above)
Comment (preactive) [reply students/preactive/Lesson3/slice.py]: fixed

File: students/preactive/Lesson3/slice.py (hunk 2 above)
Comment (preactive) [reply students/preactive/Lesson3/slice.py]: Moved asserts into name/main and refactored.

**Pull request merged**